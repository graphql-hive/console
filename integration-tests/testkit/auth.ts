import { AuthenticationClient, TokenResponse } from 'auth0';
import { ensureEnv } from './env';

const authenticationApi = new AuthenticationClient({
  domain: ensureEnv('AUTH0_DOMAIN'),
  clientId: ensureEnv('AUTH0_CLIENT_ID'),
  clientSecret: ensureEnv('AUTH0_CLIENT_SECRET'),
});

type UserID = 'main' | 'extra';
const password = ensureEnv('AUTH0_USER_PASSWORD');

const userEmails: Record<UserID, string> = {
  main: ensureEnv('AUTH0_USER_MAIN_EMAIL'),
  extra: ensureEnv('AUTH0_USER_EXTRA_EMAIL'),
};

const tokenResponsePromise: Record<UserID, Promise<TokenResponse> | null> = {
  main: null,
  extra: null,
};

export function authenticate(userId: UserID) {
  if (!tokenResponsePromise[userId]) {
    tokenResponsePromise[userId] = authenticationApi.passwordGrant({
      username: userEmails[userId],
      password,
      audience: `https://${ensureEnv('AUTH0_DOMAIN')}/api/v2/`,
      scope: 'openid profile email offline_access',
      realm: 'Username-Password-Authentication',
    });
  }

  return tokenResponsePromise[userId]!;
}
