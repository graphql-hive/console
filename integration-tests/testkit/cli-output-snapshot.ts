import stripAnsi from 'strip-ansi';
import type { SnapshotSerializer } from 'vitest';
import { ExecaError } from '@esm2cjs/execa';

// ------------------------------
// Value
// ------------------------------

type Value = ValueSansCleanable | Cleanable<ValueSansCleanable>;

type ValueSansCleanable = ValueUnwrapped | PromiseSettledResult<ValueUnwrapped>;

type ValueUnwrapped = string | ExecaError;

// ------------------------------
// Cleanable
// ------------------------------

interface Cleanable<$Value extends ValueSansCleanable> {
  value: $Value;
  clean: Cleaner;
}

type Cleaner = RegExp | ((value: string) => string);

const isCleanable = (value: unknown): value is Cleanable<ValueSansCleanable> => {
  return (
    typeof value === 'object' &&
    value !== null &&
    'value' in value &&
    'clean' in value &&
    (typeof value.clean === 'function' || value.clean instanceof RegExp)
  );
};

const applyValueCleaners = (cleaners: Cleaner[], value: string) => {
  const regexpMatchReplacer = (_: string, ...args: any[]) => {
    const captures = args.slice(0, -2);
    if (captures.length === 0) return '__VAR__';
    if (captures.length === 1) return `${captures[0]}__VAR__`;
    if (captures.length === 2) return `${captures[0]}__VAR__${captures[1]}`;
    console.log(
      'Warning: More than 2 captures in Value Cleaner given to Vitest CLI Output Serializer.',
    );
    return '__VAR__';
  };

  return cleaners.reduce((value, cleaner) => {
    if (cleaner instanceof RegExp) {
      return value.replaceAll(cleaner, regexpMatchReplacer);
    }
    return cleaner(value);
  }, value);
};

// ------------------------------
// Serializer
// ------------------------------

const test = (value: unknown): boolean => {
  if (typeof value === 'string') {
    return variableReplacements.some(replacement => replacement.pattern.test(value));
  }
  if (isCleanable(value)) {
    return test(value.value);
  }
  if (isPromiseSettledFulfilled(value)) {
    return test(value.value);
  }
  if (isPromiseSettledRejected(value)) {
    return test(value.reason);
  }
  return isExecaError(value);
};

/**
 * @remarks The `value` parameter is only type-safe by an implementation of
 * the `test` function which only returns true for matching types of values.
 */
const serialize = (value: Value): string => {
  const text = serialize_({
    value,
    valueCleaners: CliOutputSnapshot.valueCleaners,
  });
  return text + newLine + endDivider;
};

const serialize_ = ({
  value,
  valueCleaners,
}: {
  value: Value;
  valueCleaners: Cleaner[];
}): string => {
  if (isCleanable(value)) {
    const text = serialize_({
      value: value.value,
      valueCleaners: [...valueCleaners, value.clean],
    });
    return text;
  }

  if (isPromiseSettledFulfilled(value)) {
    let text = serialize_({ value: value.value, valueCleaners }) + newLine;
    text += sectionDivider('promise') + newLine;
    text += 'fulfilled';
    return text;
  }

  if (isPromiseSettledRejected(value)) {
    let text = serialize_({ value: value.reason, valueCleaners }) + newLine;
    text += sectionDivider('promise') + newLine;
    text += 'rejected';
    return text;
  }

  //
  // Scalar Cases (non-recursing)
  //
  // When running cleaners, run generic one first so that value cleaners
  // do not have to worry about ANSI codes.
  //

  if (typeof value === 'string') {
    let text = '';
    text += heading('CLI SUCCESS OUTPUT') + newLine;

    text += sectionDivider('stdout') + newLine;
    text += applyValueCleaners(valueCleaners, generalClean(value));

    return text;
  }

  if (isExecaError(value)) {
    let text = '';
    text += heading('CLI FAILURE OUTPUT') + newLine;

    text += sectionDivider('exitCode') + newLine;
    text += value.exitCode + newLine;

    text += sectionDivider('stderr') + newLine;
    text += applyValueCleaners(valueCleaners, generalClean(value.stderr || '__NONE__')) + newLine;

    text += sectionDivider('stdout') + newLine;
    text += applyValueCleaners(valueCleaners, generalClean(value.stdout || '__NONE__'));

    return text;
  }

  return String(applyValueCleaners(valueCleaners, generalClean(value)));
};

/**
 * Strip ANSI codes and mask generally known variables.
 */
const generalClean = (value: string) => {
  // We strip ANSI codes because their output can vary by platform (e.g. between macOS and GH CI linux-based runner)
  // and we don't care enough about CLI output styling to fork our snapshots for it.
  value = stripAnsi(value);
  for (const replacement of variableReplacements) {
    value = value.replaceAll(replacement.pattern, replacement.mask);
  }
  return value;
};

const variableReplacements = [
  {
    pattern: /("reference": "|"requestId": "|"https?:\/\/)[^"]+/gi,
    mask: '$1__ID__',
  },
  {
    pattern: /"https?:\/\/[^"]+/gi,
    mask: '"__URL__',
  },
  {
    pattern: /(Reference: )[^ ]+/gi,
    mask: '$1__ID__',
  },
  {
    pattern: /(https?:\/\/)[^\n ]+/gi,
    mask: '$1__URL__',
  },
];

// ------------------------------
// Type Guards
// ------------------------------

/**
 * The esm2cjs execa package we are using is not exporting the error class, so use this.
 */
const isExecaError = (value: unknown): value is ExecaError => {
  // @ts-expect-error
  return typeof value.exitCode === 'number';
};

const isPromiseSettledRejected = (value: unknown): value is PromiseRejectedResult => {
  return (
    typeof value === 'object' && value !== null && 'status' in value && value.status === 'rejected'
  );
};

const isPromiseSettledFulfilled = (value: unknown): value is PromiseFulfilledResult<unknown> => {
  return (
    typeof value === 'object' && value !== null && 'status' in value && value.status === 'fulfilled'
  );
};

// ------------------------------
// Text Helpers
// ------------------------------

const width = 50;
const newLine = '\n';

const heading = (title: string) => {
  const char = ':';
  const gap = ' ';
  const borderSize = Math.max(0, width - title.length - gap.length * 2);
  const borderSizeHalf = borderSize ? Math.round(borderSize / 2) : 0;
  const borderHalf = char.repeat(borderSizeHalf);
  return `${borderHalf}${gap}${title}${gap}${borderHalf}`;
};

const sectionDivider = (title: string) => {
  const char = '-';
  const charEnd = ':';
  const dividerSize = Math.max(0, width - title.length - 1);
  const divider = char.repeat(dividerSize) + charEnd;
  return `${title}${divider}`;
};

const endDivider = ':'.repeat(width);

// ------------------------------
// Generic Helpers
// ------------------------------

const identity = <T>(value: T): T => value;

// ------------------------------
// Interface
// ------------------------------

export namespace CliOutputSnapshot {
  export const serializer: SnapshotSerializer = {
    test,
    serialize,
  };

  /**
   * A list of cleaners that will be applied to all values before they are snapshotted.
   *
   * Any cleaners you add here will apply to all snapshots in your test file.
   *
   * The value will be cleaned with the general cleaner first. So for example ANSI codes will have been stripped already.
   *
   * A cleaner can be a RegExp instance or an arbitrary function that receives and returns a string.
   *
   * #### RegExp Cleaners
   * When using RegExps, note the following.
   *
   * With regards to captured groups:
   *
   * String method `.replaceAll` is used so your RegExp MUST be using `g` flag.
   *
   * With regards to flags:
   * - 0 captures: Replace with `__VAR__`
   * - 1 capture: Replace with `$1__VAR__`
   * - 2 captures: Replace with `$1__VAR__$2`
   * - 3+ captures: Replace with `__VAR__` and log a warning. (Don't do this.)
   */
  export const valueCleaners: Cleaner[] = [];

  /**
   * Wrap a value with a cleaner for just that value.
   *
   * For details about cleaners, see {@link valueCleaners}.
   */
  export const withCleaner = (
    value: ValueSansCleanable,
    valueClean: Cleaner,
  ): Cleanable<ValueSansCleanable> => {
    return {
      value,
      clean: valueClean,
    };
  };
}
