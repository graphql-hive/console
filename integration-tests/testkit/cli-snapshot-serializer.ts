import stripAnsi from 'strip-ansi';
import type { SnapshotSerializer } from 'vitest';
import { ExecaError } from '@esm2cjs/execa';

const test = (value: unknown): boolean => {
  if (typeof value === 'string') {
    return variableReplacements.some(replacement => replacement.pattern.test(value));
  }
  if (isPromiseSettledFulfilled(value)) {
    return test(value.value);
  }
  if (isPromiseSettledRejected(value)) {
    return test(value.reason);
  }
  return isExecaError(value);
};

type Value = string | ExecaError | PromiseSettledResult<string | ExecaError>;

/**
 * @remarks The `value` parameter is only type-safe by an implementation of
 * the `test` function which only returns true for these types of values.
 */
const serialize = (value: Value): string => {
  const text = serialize_(value);
  return text + newLine + endDivider;
};

const serialize_ = (value: Value): string => {
  if (isPromiseSettledFulfilled(value)) {
    return serialize_(value.value) + newLine + sectionDivider('promise') + newLine + 'fulfilled';
  }

  if (isPromiseSettledRejected(value)) {
    return serialize_(value.reason) + newLine + sectionDivider('promise') + newLine + 'rejected';
  }

  if (typeof value === 'string') {
    let text = '';
    text += heading('CLI SUCCESS OUTPUT') + newLine;

    text += sectionDivider('stdout') + newLine;
    text += clean(value);

    return text;
  }

  if (isExecaError(value)) {
    let text = '';
    text += heading('CLI FAILURE OUTPUT') + newLine;

    text += sectionDivider('exitCode') + newLine;
    text += value.exitCode + newLine;

    text += sectionDivider('stderr') + newLine;
    text += clean(value.stderr || '__NONE__') + newLine;

    text += sectionDivider('stdout') + newLine;
    text += clean(value.stdout || '__NONE__');

    return text;
  }

  return String(value);
};

export const cliOutputSnapshotSerializer: SnapshotSerializer = {
  test,
  serialize,
};

const variableReplacements = [
  {
    pattern: /("reference": "|"requestId": "|"https?:\/\/)[^"]+/gi,
    mask: '$1__ID__',
  },
  {
    pattern: /"https?:\/\/[^"]+/gi,
    mask: '"__URL__',
  },
  {
    pattern: /(Reference: )[^ ]+/gi,
    mask: '$1__ID__',
  },
  {
    pattern: /(https?:\/\/)[^\n ]+/gi,
    mask: '$1__URL__',
  },
];

/**
 * Strip ANSI codes and mask variables.
 */
const clean = (value: string) => {
  // We strip ANSI codes because their output can vary by platform (e.g. between macOS and GH CI linux-based runner)
  // and we don't care enough about CLI output styling to fork our snapshots for it.
  value = stripAnsi(value);
  for (const replacement of variableReplacements) {
    value = value.replaceAll(replacement.pattern, replacement.mask);
  }
  return value;
};

/**
 * The esm2cjs execa package we are using is not exporting the error class, so use this.
 */
const isExecaError = (value: unknown): value is ExecaError => {
  // @ts-expect-error
  return typeof value.exitCode === 'number';
};

const isPromiseSettledRejected = (value: unknown): value is PromiseRejectedResult => {
  return (
    typeof value === 'object' && value !== null && 'status' in value && value.status === 'rejected'
  );
};

const isPromiseSettledFulfilled = (value: unknown): value is PromiseFulfilledResult<unknown> => {
  return (
    typeof value === 'object' && value !== null && 'status' in value && value.status === 'fulfilled'
  );
};

const width = 50;
const newLine = '\n';

const heading = (title: string) => {
  const char = ':';
  const gap = ' ';
  const borderSize = Math.max(0, width - title.length - gap.length * 2);
  const borderSizeHalf = borderSize ? Math.round(borderSize / 2) : 0;
  const borderHalf = char.repeat(borderSizeHalf);
  return `${borderHalf}${gap}${title}${gap}${borderHalf}`;
};

const sectionDivider = (title: string) => {
  const char = '-';
  const charEnd = ':';
  const dividerSize = Math.max(0, width - title.length - 1);
  const divider = char.repeat(dividerSize) + charEnd;
  return `${title}${divider}`;
};

const endDivider = ':'.repeat(width);
