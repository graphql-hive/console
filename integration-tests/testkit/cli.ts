import * as utils from 'dockest/test-helper';
import { run } from '../../packages/libraries/cli/src/index';

const registryAddress = utils.getServiceAddress('server', 3001);

export async function schemaPublish(args: string[]) {
  return run(['schema:publish', `--registry`, `http://${registryAddress}/graphql`, ...args]);
}

export async function schemaCheck(args: string[]) {
  return run(['schema:check', `--registry`, `http://${registryAddress}/graphql`, ...args]);
}
