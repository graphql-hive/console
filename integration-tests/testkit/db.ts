import { sql, DatabasePoolConnectionType } from 'slonik';

export const resetDb = async (conn: DatabasePoolConnectionType) => {
  const migrationTables = ['migrations'];

  const result = await conn.many<{ tablename: string }>(sql`
    SELECT "tablename"
    FROM "pg_tables"
    WHERE "schemaname" = 'public';
  `);

  const tablenames = result.map(({ tablename }) => tablename).filter(tablename => !migrationTables.includes(tablename));

  if (tablenames.length) {
    await conn.query(sql`
      TRUNCATE TABLE 
        ${sql.join(
          tablenames.map(name => sql.identifier([name])),
          sql`,`
        )}
        RESTART IDENTITY
      ;
    `);
  }
};
