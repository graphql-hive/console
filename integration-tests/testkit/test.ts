/**
 * This module uses Vitest's fixture system to make common usage patterns
 * of our testkit easily consumable in test cases. @see https://vitest.dev/guide/test-context.html#test-extend
 */

import { test as testBase } from 'vitest';
import { CLI, createCLI } from './cli';
import { ProjectType } from './gql/graphql';
import { initSeed, OrgSeed, OwnerSeed, ProjectSeed, Seed, TargetAccessTokenSeed } from './seed';

interface Context {
  seed: Seed;
  owner: OwnerSeed;
  org: OrgSeed;
  //
  // "single" branch
  //
  projectSingle: ProjectSeed;
  tokenForProjectSingle: TargetAccessTokenSeed;
  cliForProjectSingle: CLI;
  //
  // "federation" branch
  //
  projectFederation: ProjectSeed;
  tokenForProjectFederation: TargetAccessTokenSeed;
  cliForProjectFederation: CLI;
  //
  // "stitching" branch
  //
  projectStitching: ProjectSeed;
  tokenForProjectStitching: TargetAccessTokenSeed;
  cliForProjectStitching: CLI;
}

export const test = testBase.extend<Context>({
  seed: async ({}, use) => {
    const seed = await initSeed();
    await use(seed);
  },
  owner: async ({ seed }, use) => {
    const owner = await seed.createOwner();
    await use(owner);
  },
  org: async ({ owner }, use) => {
    const org = await owner.createOrg();
    await use(org);
  },
  //
  // "single" branch
  //
  projectSingle: async ({ org }, use) => {
    const project = await org.createProject(ProjectType.Single);
    await use(project);
  },
  tokenForProjectSingle: async ({ projectSingle }, use) => {
    const token = await projectSingle.createTargetAccessToken({});
    await use(token);
  },
  cliForProjectSingle: async ({ tokenForProjectSingle }, use) => {
    const cli = createCLI({
      readwrite: tokenForProjectSingle.secret,
      readonly: tokenForProjectSingle.secret,
    });
    await use(cli);
  },
  //
  // "federation" branch
  //
  projectFederation: async ({ org }, use) => {
    const project = await org.createProject(ProjectType.Federation);
    await use(project);
  },
  tokenForProjectFederation: async ({ projectFederation }, use) => {
    const token = await projectFederation.createTargetAccessToken({});
    await use(token);
  },
  cliForProjectFederation: async ({ tokenForProjectFederation }, use) => {
    const cli = createCLI({
      readwrite: tokenForProjectFederation.secret,
      readonly: tokenForProjectFederation.secret,
    });
    await use(cli);
  },
  //
  // "stitching" branch
  //
  projectStitching: async ({ org }, use) => {
    const project = await org.createProject(ProjectType.Stitching);
    await use(project);
  },
  tokenForProjectStitching: async ({ projectStitching }, use) => {
    const token = await projectStitching.createTargetAccessToken({});
    await use(token);
  },
  cliForProjectStitching: async ({ tokenForProjectStitching }, use) => {
    const cli = createCLI({
      readwrite: tokenForProjectStitching.secret,
      readonly: tokenForProjectStitching.secret,
    });
    await use(cli);
  },
});
