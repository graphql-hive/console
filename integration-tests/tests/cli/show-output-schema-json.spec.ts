import { exec } from '../../testkit/cli';
import { test } from '../../testkit/test';
import { SnapshotSerializers } from './__snapshot_serializers__/__';

expect.addSnapshotSerializer(SnapshotSerializers.cliOutput);

interface TestCase {
  command:
    | 'whoami'
    | 'schema:publish'
    | 'schema:check'
    | 'schema:delete'
    | 'schema:fetch'
    | 'app:create'
    | 'app:publish';
  args?: Record<string, string>;
}

// prettier-ignore
const testCases: TestCase[] = [
  { command: 'whoami' },
  { command: 'schema:publish' },
  { command: 'schema:check' },
  { command: 'schema:delete' },
  { command: 'schema:fetch' },
  { command: 'app:create' },
  { command: 'app:publish' },
];

test.each(testCases)('--show-output-schema-json - %s', async ({ command, args }) => {
  const preparedArgs = args
    ? Object.entries(args)
        .map(([key, value]) => `--${key}=${value}`)
        .join(' ')
    : '';
  const preparedCommand = `${command} ${preparedArgs} --show-output-schema-json`;

  const [outputText, outputJson] = await Promise.all([
    exec(preparedCommand),
    exec(`${preparedCommand} --json`),
  ]);
  expect(outputText).toMatch(outputJson); // --json flag does not change the output
  const schema = JSON.parse(outputText);
  await expect(outputText).toMatchSnapshot();
  await expect(JSON.stringify(schema, null, 2)).toMatchFileSnapshot(
    `./__snapshots__/show-output-schema-json/${command.replace(':', '__')}.json`,
  );
});
