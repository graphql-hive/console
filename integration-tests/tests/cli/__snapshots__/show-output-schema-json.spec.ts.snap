// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`--show-output-schema-json - { command: 'app:create' } 1`] = `{"anyOf":[{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessSkipAppCreate","const":"SuccessSkipAppCreate","type":"string"},"status":{"anyOf":[{"const":"active","type":"string"},{"const":"pending","type":"string"},{"const":"retired","type":"string"}]}},"required":["type","status"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessAppCreate","const":"SuccessAppCreate","type":"string"},"id":{"minLength":1,"type":"string"},"operationsCount":{"type":"number"}},"required":["type","id","operationsCount"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureAppCreate","const":"FailureAppCreate","type":"string"},"message":{"type":"string"}},"required":["type","message"]}},"required":["type","reference","suggestions","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureInvalidManifestModel","const":"FailureInvalidManifestModel","type":"string"},"errors":{"type":"array","items":{"$id":"MaterializedValueError","type":"object","properties":{"message":{"type":"string"},"path":{"type":"string"},"value":{},"errors":{"type":"array","items":{"$ref":"MaterializedValueError"}}},"required":["message","path","value","errors"]}}},"required":["type","errors"]}},"required":["type","reference","suggestions","data"]}]}`;

exports[`--show-output-schema-json - { command: 'app:publish' } 1`] = `{"anyOf":[{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessSkipAppPublish","const":"SuccessSkipAppPublish","type":"string"},"name":{"minLength":1,"type":"string"},"version":{"minLength":1,"type":"string"}},"required":["type","name","version"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessAppPublish","const":"SuccessAppPublish","type":"string"},"name":{"minLength":1,"type":"string"},"version":{"minLength":1,"type":"string"}},"required":["type","name","version"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureAppPublish","const":"FailureAppPublish","type":"string"},"message":{"type":"string"}},"required":["type","message"]}},"required":["type","reference","suggestions","data"]}]}`;

exports[`--show-output-schema-json - { command: 'schema:check' } 1`] = `{"anyOf":[{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessSchemaCheck","const":"SuccessSchemaCheck","type":"string"},"diffType":{"anyOf":[{"const":"initial","type":"string"},{"const":"change","type":"string"},{"const":"unknown","type":"string"}]},"changes":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"},"criticality":{"anyOf":[{"const":"Breaking","type":"string"},{"const":"Dangerous","type":"string"},{"const":"Safe","type":"string"}]},"isSafeBasedOnUsage":{"type":"boolean"},"approval":{"anyOf":[{"type":"object","properties":{"by":{"anyOf":[{"type":"object","properties":{"displayName":{"anyOf":[{"type":"string"},{"type":"null"}]}},"required":["displayName"]},{"type":"null"}]}},"required":["by"]},{"type":"null"}]}},"required":["message","criticality","isSafeBasedOnUsage","approval"]}},"warnings":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"},"source":{"anyOf":[{"type":"string"},{"type":"null"}]},"line":{"anyOf":[{"type":"number"},{"type":"null"}]},"column":{"anyOf":[{"type":"number"},{"type":"null"}]}},"required":["message","source","line","column"]}},"url":{"anyOf":[{"format":"uri","type":"string"},{"type":"null"}]}},"required":["type","diffType","changes","warnings","url"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessSchemaCheckGitHub","const":"SuccessSchemaCheckGitHub","type":"string"},"message":{"type":"string"}},"required":["type","message"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaCheck","const":"FailureSchemaCheck","type":"string"},"changes":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"},"criticality":{"anyOf":[{"const":"Breaking","type":"string"},{"const":"Dangerous","type":"string"},{"const":"Safe","type":"string"}]},"isSafeBasedOnUsage":{"type":"boolean"},"approval":{"anyOf":[{"type":"object","properties":{"by":{"anyOf":[{"type":"object","properties":{"displayName":{"anyOf":[{"type":"string"},{"type":"null"}]}},"required":["displayName"]},{"type":"null"}]}},"required":["by"]},{"type":"null"}]}},"required":["message","criticality","isSafeBasedOnUsage","approval"]}},"warnings":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"},"source":{"anyOf":[{"type":"string"},{"type":"null"}]},"line":{"anyOf":[{"type":"number"},{"type":"null"}]},"column":{"anyOf":[{"type":"number"},{"type":"null"}]}},"required":["message","source","line","column"]}},"errors":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"}},"required":["message"]}},"url":{"anyOf":[{"format":"uri","type":"string"},{"type":"null"}]}},"required":["type","changes","warnings","errors","url"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaCheckGitHub","const":"FailureSchemaCheckGitHub","type":"string"},"message":{"type":"string"}},"required":["type","message"]}},"required":["type","reference","suggestions","data"]}]}`;

exports[`--show-output-schema-json - { command: 'schema:delete' } 1`] = `{"anyOf":[{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessSchemaDelete","const":"SuccessSchemaDelete","type":"string"}},"required":["type"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaDelete","const":"FailureSchemaDelete","type":"string"},"errors":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"}},"required":["message"]}}},"required":["type","errors"]}},"required":["type","reference","suggestions","data"]}]}`;

exports[`--show-output-schema-json - { command: 'schema:fetch' } 1`] = `{"anyOf":[{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaFetchMissingSchema","const":"FailureSchemaFetchMissingSchema","type":"string"}},"required":["type"]}},"required":["type","reference","suggestions","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaFetchInvalidSchema","const":"FailureSchemaFetchInvalidSchema","type":"string"}},"required":["type"]}},"required":["type","reference","suggestions","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaFetchMissingSDLType","const":"FailureSchemaFetchMissingSDLType","type":"string"}},"required":["type"]}},"required":["type","reference","suggestions","data"]},{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessOutputFile","const":"SuccessOutputFile","type":"string"},"path":{"type":"string"},"bytes":{"type":"number"}},"required":["type","path","bytes"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessOutputStdout","const":"SuccessOutputStdout","type":"string"},"content":{"type":"string"}},"required":["type","content"]}},"required":["type","data"]}]}`;

exports[`--show-output-schema-json - { command: 'schema:publish' } 1`] = `{"anyOf":[{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessSchemaPublish","const":"SuccessSchemaPublish","type":"string"},"diffType":{"anyOf":[{"const":"initial","type":"string"},{"const":"change","type":"string"},{"const":"unknown","type":"string"}]},"message":{"anyOf":[{"type":"string"},{"type":"null"}]},"changes":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"},"criticality":{"anyOf":[{"const":"Breaking","type":"string"},{"const":"Dangerous","type":"string"},{"const":"Safe","type":"string"}]},"isSafeBasedOnUsage":{"type":"boolean"},"approval":{"anyOf":[{"type":"object","properties":{"by":{"anyOf":[{"type":"object","properties":{"displayName":{"anyOf":[{"type":"string"},{"type":"null"}]}},"required":["displayName"]},{"type":"null"}]}},"required":["by"]},{"type":"null"}]}},"required":["message","criticality","isSafeBasedOnUsage","approval"]}},"url":{"anyOf":[{"format":"uri","type":"string"},{"type":"null"}]}},"required":["type","diffType","message","changes","url"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaPublish","const":"FailureSchemaPublish","type":"string"},"changes":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"},"criticality":{"anyOf":[{"const":"Breaking","type":"string"},{"const":"Dangerous","type":"string"},{"const":"Safe","type":"string"}]},"isSafeBasedOnUsage":{"type":"boolean"},"approval":{"anyOf":[{"type":"object","properties":{"by":{"anyOf":[{"type":"object","properties":{"displayName":{"anyOf":[{"type":"string"},{"type":"null"}]}},"required":["displayName"]},{"type":"null"}]}},"required":["by"]},{"type":"null"}]}},"required":["message","criticality","isSafeBasedOnUsage","approval"]}},"errors":{"type":"array","items":{"type":"object","properties":{"message":{"type":"string"}},"required":["message"]}},"url":{"anyOf":[{"format":"uri","type":"string"},{"type":"null"}]}},"required":["type","changes","errors","url"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessSchemaPublishGitHub","const":"SuccessSchemaPublishGitHub","type":"string"},"message":{"type":"string"}},"required":["type","message"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaPublishGitHub","const":"FailureSchemaPublishGitHub","type":"string"},"message":{"type":"string"}},"required":["type","message"]}},"required":["type","reference","suggestions","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureSchemaPublishInvalidGraphQLSchema","const":"FailureSchemaPublishInvalidGraphQLSchema","type":"string"},"message":{"type":"string"},"locations":{"type":"array","items":{"type":"object","properties":{"line":{"type":"number"},"column":{"type":"number"}},"required":["line","column"]}}},"required":["type","message","locations"]}},"required":["type","reference","suggestions","data"]}]}`;

exports[`--show-output-schema-json - { command: 'whoami' } 1`] = `{"anyOf":[{"type":"object","properties":{"type":{"default":"success","const":"success","type":"string"},"data":{"type":"object","properties":{"type":{"default":"SuccessWhoami","const":"SuccessWhoami","type":"string"},"token":{"type":"object","properties":{"name":{"type":"string"}},"required":["name"]},"organization":{"type":"object","properties":{"slug":{"type":"string"},"url":{"format":"uri","type":"string"}},"required":["slug","url"]},"project":{"type":"object","properties":{"type":{"type":"string"},"slug":{"type":"string"},"url":{"format":"uri","type":"string"}},"required":["type","slug","url"]},"target":{"type":"object","properties":{"slug":{"type":"string"},"url":{"format":"uri","type":"string"}},"required":["slug","url"]},"authorization":{"type":"object","properties":{"schema":{"type":"object","properties":{"publish":{"type":"boolean"},"check":{"type":"boolean"}},"required":["publish","check"]}},"required":["schema"]}},"required":["type","token","organization","project","target","authorization"]}},"required":["type","data"]},{"type":"object","properties":{"type":{"const":"failure","type":"string"},"reference":{"anyOf":[{"type":"string"},{"type":"null"}]},"suggestions":{"type":"array","items":{"type":"string"}},"data":{"type":"object","properties":{"type":{"default":"FailureWhoamiTokenNotFound","const":"FailureWhoamiTokenNotFound","type":"string"},"message":{"type":"string"}},"required":["type","message"]}},"required":["type","reference","suggestions","data"]}]}`;
